### ISSUE TITLE
Integrate sweet.js with ES6 modules

### ISSUE THREAD

**Issue by disnet** [sweet-js/sweet-core#233]
There are two papers on macros and modules that are probably relevant. [Composable and Compilable Macros](http://www.cs.utah.edu/plt/publications/macromod.pdf) and [Implicit Phasing for R6RS Libraries](http://www.cs.indiana.edu/~dyb/pubs/implicit-phasing.pdf).

The basic idea of both papers is that you have to track the phase level any bit of syntax currently is in. The phase isn't as simple as compile-time vs. run-time because a case macro (compile-time) might need to import and use another macro (earlier compile-time) which might also need to use a macro (even _earlier_ compile-time) and so on. 

``` js
// bar.js
export macro bar {
    case {_ ()} => {
        // ...
    }
}
```

``` js
// foo.js
import { bar } from "bar"
export macro foo {
    case {_ () } => { 
        bar()
        // ... 
    }
}
```

``` js
// main.js
import { foo } from "foo"
foo()
```

The papers outline two approaches we could take. The first, following the "Composable and Compilable" paper and Racket, involves explicitly declaring what phase level an import is for. 

``` js
import { $ } from "jquery"
import forSyntax { _ } from "underscore"

let m = macro {
    case { } => {
        $("#foo") // error: $ not bound
        _.map([1,2,3], function (x) { ... });
        // ...
    }
}
$("#foo") // works fine
_.map(...) // error: _ is not bound
```

This gets ugly though when you have a macro definition in your macro definition since you need to import up two phase levels:

``` js
import { $ } from "jquery"
import forSyntax { _ } from "underscore"
import forSyntax forSyntax { ld } from "lodash"
// or maybe following racket:
import meta 2 { ld } from "lodash"

let m = macro {
    case { } => {
        let n = macro {
            case {} => {
                ld.map(...)
                // ...
            }
        }
        $("#foo") // error: $ not bound
        _.map([1,2,3], function (x) { ... });
        // ...
    }
}
$("#foo") // works fine
_.map(...) // error: _ is not bound
```

The "Implicit Phasing" paper gives a technique for tracking all the phasing implicitly so you wouldn't need to do `import forSyntax`. The paper seems to suggest that there are no real trade offs in doing this but I seem to remember @dherman mention there was a good reason Racket stayed with explicitly phase labeling (something about the REPL maybe?). 

I would prefer to keep the phasing implicit if possible since forcing macro authors to think about the details of module phasing seems not great.

**Comment by elibarzilay** [comment]
Some general random comments:

* One of the major surprises of implicit phases and how it leads to "lazy loading": a practical result of that is that you can't really use toplevel non-definition expressions for side-effects.  That's because you can't just require some library to execute these expressions.  IIRC, the Ikarus approach was to have an `init` function for such things.  Or something.

* This has been a very long debate in R6RS.  Personally, I find no problem with the explicit approach (but I'm probably biased too), but when I realized the above that seemed like a major problem which made my preference much stronger.  If ES6 modules are expected to be executed when required, then it makes choosing implicit phases much more impractical.

* Re the multiple instances vs single ones: I can only say that Racket's module system seemed completely crazy to be doing that.  But the results -- the ability to write any code at any level -- is something that you shouldn't underestimate.  I didn't even realize how significant it was until I had a discussion with a random schemer who claimed that doing "real things" at the macro level, like running a full parser, or invoking an external compiler, or fetching some resource from the web, all of these sounded completely crazy to that person.  (And obviously, no -- you cannot just be careful -- since state is there in many unexpected cases, like creating a new type.)

* The problem with multiple instances -- cases where you want a single instance -- are all cases where you're talking about low-level stuff like buffered IO, a single DB handle, etc.  That is, cases where you're talking to the external world and need to face inherent state phase leakages.  In practice there are few of these, and they are better dealt with at the same level you're dealing with foreign code.

* So my subjective conclusion from all of this is that Racket leans towards the choices it made because (a) it views all languages as equal and implementable, and therefore it never wants to confuse an `if` at one level with an `if` at another (I vaguely remember stuff like this turning extremely confusing with an implicit system); (b) there's generally a decision to allow any tools at the macro level, and therefore multiple instances make much more sense.

**Comment by disnet** [comment]
Yep, I was long ago convinced the explicit/multiple instances approach is the right way to go for Sweet. Currently hacking on it as a matter of fact :)

Thanks for the additional insight!

**Comment by elibarzilay** [comment]
@disnet: I'm happy to hear that it was preaching to the choir!  It would be interesting to know how you'd implement multiple instances of the same library -- looking at some loaders I got the general impression that it *could* be easy since they're basically doing it in the form of one big function.

Also, having modules that provide/require syntax in JS would be amazing IMO.  Getting macros is one thing, but getting how to do them with a module system is a whole new level -- and that's exactly the thing that Racket did earlier than all other Schemes, eventually making it a language for implementing PLs.

**Comment by isiahmeadows** [comment]
What's the status of this?

**Comment by disnet** [comment]
@isiahmeadows we have basic [support implemented](http://sweetjs.org/doc/1.0/tutorial.html#_sweet_modules) for importing a macro along with importing `for syntax`.

It's not full support by any means. It's only one phase level (no recursive imports) and only the `import { named }` form is supported.

I'm currently hacking on it though so full support will be coming Real Soon™️.

**Comment by isiahmeadows** [comment]
Yay! :-)
